crates/macros/src/lib.rs

```diff
@@ -35,20 +61,59 @@ fn impl_block_data(ast: &DeriveInput) -> Result<TokenStream> {
         ));
     }
 
-    let gen = quote! {
+    Ok(fields)
+}
+
+fn impl_block_data(name: &Ident, fields: &Vec<Field>) -> TokenStream2 {
+    let idents: Vec<Ident> = fields.iter().map(|f| f.to_owned().ident.unwrap()).collect();
+    quote!(
         impl BlockData for #name {
-            fn from_raw(raw: &RawBlockStateProperties) -> Option<Self>
+            fn from_raw(raw: &RawBlockStateProperties, valid: ValidProperties) -> Option<Self>
             where
                 Self: Sized,
             {
-                Some(Self { #(#fields: raw.#fields?),*, })
+                Some(Self { #(#idents: raw.#idents?),*,valid_properties: valid, })
             }
 
             fn apply(&self, raw: &mut RawBlockStateProperties) {
-                #(raw.#fields.replace(self.#fields));*;
+                #(raw.#idents.replace(self.#idents));*;
             }
         }
-    };
+    )
+}
 
-    Ok(gen.into())
+fn impl_getters_and_setters(name: &Ident, fields: &Vec<Field>) -> TokenStream2 {
+    let mut getters_setters = Vec::new();
+    for field in fields {
+        let field = field.to_owned();
+        let ident = field.ident.unwrap();
+        let ty = field.ty;
+        let get_ident = format_ident!("get_{}", ident);
```

Comment:
Naming convention in Rust typically removes the `get_` prefix for getters, so e.g. `get_age()` would just be `age()`.

Reply:
Is the setter fine as it is?

And I don't really like the `valid_` function name either but couldn't come up with any better alternatives.

Reply:
Yes, the setter is correct—set_{property}` follows the naming conventions.

For the `valid` function, it would be ideal to change the property name to plural, like `valid_ages`. Not sure how difficult this would be to handle in the derive macro, though.

Reply:
Well I think I could possibly set the `s` at the end of the `Ident` if it doesn't already contain one (like `bamboo_leaves`). But I'm not sure if it would be better.

Reply:
That should work fine for most properties. We can special-case properties with different plural forms, I guess.